code/datums/map_generators/map_placer.dm

```diff
@@ -0,0 +1,332 @@
+
+	var/current_run = GENERATE_STAGE_BUILD_CACHE_START
+	var/run_stage = 0
+
+	var/list/area_cache = list()
+	var/list/model_cache
+	var/space_key = null
+	var/list/bounds
+
+	//=========================
+	// Build Cache Locals
+	//=========================
+
+	var/list/grid_models
+
+	var/grid_model_index
+
+	var/model_key
+	var/model
+	var/list/members
+	var/list/members_attributes
+	var/index
+	var/old_position
+	var/dpos
+
+	//=========================
+	// Build Coordinate Locals
+	//=========================
+
+	var/datum/grid_set/gset
+
+	/// The index of the grid line we are currently working on.
+	/// Start at infinity as we need to access the outer loop first.
+	var/current_grid_line = INFINITY
+
+	var/ycrd
+	var/zcrd
+	var/zexpansion
+
+/datum/map_generator/map_place/New(datum/parsed_map/placing_template, x_offset = 1, y_offset = 1, z_offset = world.maxz + 1, cropMap = FALSE, no_changeturf = FALSE, x_lower = -INFINITY, x_upper = INFINITY, y_lower = -INFINITY, y_upper = INFINITY, placeOnTop = FALSE)
+	. = ..()
+	src.placing_template = placing_template
+	src.x_offset = x_offset
+	src.y_offset = y_offset
+	src.z_offset = z_offset
+	crop_map = cropMap
+	src.no_changeturf = no_changeturf
+	src.x_lower = x_lower
+	src.x_upper = x_upper
+	src.y_lower = y_lower
+	src.y_upper = y_upper
+	place_on_top = placeOnTop
+
+/datum/map_generator/map_place/execute_run()
+	..()
+	if (current_run == GENERATE_STAGE_BUILD_CACHE_START)
+		build_cache_start()
+	if (current_run == GENERATE_STAGE_BUILD_CACHE)
+		build_cache()
+	if (current_run == GENERATE_STAGE_BUILD_COORDINATES_START)
+		build_coordinates_start()
+	if (current_run == GENERATE_STAGE_BUILD_COORDINATES)
+		build_coordinates()
+	. = current_run == GENERATE_STAGE_COMPLETED
+
+/datum/map_generator/map_place/proc/set_stage(stage)
+	run_stage = 1
+	current_run = stage
+
+//======================================
+// COORDINATE BUILDING
+//======================================
+
+/datum/map_generator/map_place/proc/build_coordinates_start()
+	//Locate the space key
+	space_key = model_cache[SPACE_KEY]
+	//Set them all to the same reference, so changing one affects the other
+	bounds = placing_template.bounds = list(1.#INF, 1.#INF, 1.#INF, -1.#INF, -1.#INF, -1.#INF)
+	//Move to the next stage
+	set_stage(GENERATE_STAGE_BUILD_COORDINATES)
+
+/datum/map_generator/map_place/proc/build_coordinates()
+	while (TRUE)
+		// Perform inner loop first
+		while (gset && current_grid_line <= length(gset.gridLines))
+			// Build a single grid line
+			build_coordinate_grid_line()
+			//Building the grid line overran tick
+			if (TICK_CHECK)
+				return
+		// Enumerate to the next grid set
+		run_stage ++
+		// Check if we are still within bounds
+		if (run_stage - 1 > length(placing_template.gridSets))
+			if(!no_changeturf)
+				for(var/t in block(locate(bounds[MAP_MINX], bounds[MAP_MINY], bounds[MAP_MINZ]), locate(bounds[MAP_MAXX], bounds[MAP_MAXY], bounds[MAP_MAXZ])))
+					var/turf/T = t
```

Comment:
you can just force the loop to have type

Reply:
I could, this is largely a copy and paste from map code rather than a rewrite